### PR TITLE
search: implement whoosh search engine

### PR DESCRIPTION
**Pull request by PascalRepond**
* Updates README.
* Also adds some cosmetic fixes.
* Adds a CLI with commands to initialize the index and index all pages.

### CODE REVIEW

File: flask_wiki/api.py [hunk 1]
```diff
@@ -167,6 +173,19 @@ def render(self):
         self.modification_datetime = datetime.fromtimestamp(
             os.path.getmtime(self.path))
 
+    def index(self):
+        """Index page data for whoosh search engine."""
+        ix = index.open_dir(current_app.config.get('WIKI_INDEX_DIR'))
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: Rename parameter `ix` (The name doesn't say it all). `index_dir` ?

File: flask_wiki/api.py [hunk 2]
```diff
@@ -230,6 +250,22 @@ def tags(self, value):
...
+        Returns the raw text of the body without markdown or html markup,
+        used for indexing and search results display.
+        """
+        html = markdown.markdown(self.body)
+        html = BeautifulSoup(html, 'html.parser')
+        return html.get_text(separator=' ')
+
+    @raw_body.setter
+    def raw_body(self, value):
+        """."""
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: docstring

File: flask_wiki/api.py [hunk 3]
```diff
@@ -309,9 +345,54 @@ def delete(self, url):
         if not self.exists(url):
             return False
         os.remove(path)
+        ix = index.open_dir(current_app.config.get('WIKI_INDEX_DIR'))
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: same here

File: flask_wiki/api.py [hunk 4]
```diff
@@ -353,12 +439,12 @@ def index_by(self, key):
 
     def get_by_title(self, title):
         """."""
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: doctring

File: flask_wiki/api.py [hunk 5]
```diff
@@ -353,12 +439,12 @@ def index_by(self, key):
...
-        pages = self.index(attr='title')
+        pages = self.list_pages(attr='title')
         return pages.get(title)
 
     def get_tags(self):
         """."""
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: doctring

File: flask_wiki/api.py [hunk 6]
```diff
@@ -374,7 +460,7 @@ def get_tags(self):
 
     def index_by_tag(self, tag):
         """."""
```
Comment (Garfield-fr) [inline flask_wiki/api.py]: doctring

File: flask_wiki/views.py [hunk 7]
```diff
@@ -243,10 +244,13 @@ def files():
 def search():
     """."""
     query = request.args.get('q', '')
-    results = current_wiki.search(query)
-    return render_template(
-        current_app.config.get('WIKI_SEARCH_TEMPLATE'),
-        results=results, query=query)
+    with current_app.app_context():
+        ix = whoosh_index.open_dir(current_app.config.get('WIKI_INDEX_DIR'))
```
Comment (Garfield-fr) [inline flask_wiki/views.py]: same here

**Review by Garfield-fr — APPROVED** [review]

**Review by rerowep — APPROVED** [review]

**Review by zannkukai — APPROVED** [review]

**Pull request merged**